patches/016-olsrd-procd.patch

```diff
@@ -0,0 +1,1886 @@
++				$IP4T -I forwarding_rule -o tnl_+ -j ACCEPT
++				$IP4T -t nat -I postrouting_rule -o tnl_+ -j MASQUERADE
++				# Allow forwarding from tunl0 to (all) wan-interfaces
++				if [ "$nowan"="0" ]; then
++					for IFACE in $wanifnames; do
++						$IP4T -A forwarding_rule -i tunl0 -o $IFACE -j ACCEPT
++					done
++				fi
++				# Allow incoming ipip on all olsr-interfaces
++				for IFACE in $ifsglobal; do
++					$IP4T -I input_rule -i $IFACE -p 4 -j ACCEPT
++				done
++			elif [ "$smartgatewayuplink" == "ipv6" ]; then
++				$IP6T -I forwarding_rule -o tnl_+ -j ACCEPT
++				if [ "$nowan"="0" ]; then
++					for IFACE in $wanifnames; do
++						$IP6T -A forwarding_rule -i tunl0 -o $IFACE -j ACCEPT
++					done
++				fi
++				for IFACE in $ifsglobal; do
++					$IP6T -I input_rule -i $IFACE -p 4 -j ACCEPT
++				done
++			else
++				$IP4T -t nat -I postrouting_rule -o tnl_+ -j MASQUERADE
++				for IPT in $IP4T $IP6T; do
++					$IPT -I forwarding_rule -o tnl_+ -j ACCEPT
++					if [ "$nowan"="0" ]; then
++						for IFACE in $wanifnames; do
++							$IPT -A forwarding_rule -i tunl0 -o $IFACE -j ACCEPT
++						done
++					fi
++					for IFACE in $ifsglobal; do
++						$IPT -I input_rule -i $IFACE -p 4 -j ACCEPT
++					done
++				done
++			fi
++		fi
++	fi
++}
++
++olsrd_generate_config() {
++	UCI_CONF_NAME="$1"
++	SYSTEM_HOSTNAME=
++	SYSTEM_LAT=
++	SYSTEM_LON=
++	config_load system
++	config_foreach system_config system
++
++	option_cb() {
++		olsrd_update_schema "option" "$@"
++	}
++
++	list_cb() {
++		olsrd_update_schema "list" "$@"
++	}
++
++	. /lib/functions/network.sh
++
++	config_load $UCI_CONF_NAME
++	reset_cb
++
++	OLSRD_CONFIG_FILE=
++	config_foreach olsrd_find_config_file olsrd
++
++	if [ -z "$OLSRD_CONFIG_FILE" ]; then
++		mkdir -p -- /var/etc/
++		olsrd_write_config $UCI_CONF_NAME > /var/etc/$UCI_CONF_NAME.conf || return 1
++		if [ "$INTERFACES_COUNT" -gt 0 -a "$OLSRD_COUNT" -gt 0 ]; then
++			OLSRD_CONFIG_FILE=/var/etc/$UCI_CONF_NAME.conf
++		fi
++	fi
++
++	[ -z "$OLSRD_CONFIG_FILE" ] && return 1
++}
+diff --git a/package/routing/olsrd/files/olsrd4.init b/package/routing/olsrd/files/olsrd4.init
+index 718b3e0..8b8f2e5 100644
+--- a/package/routing/olsrd/files/olsrd4.init
++++ b/package/routing/olsrd/files/olsrd4.init
+@@ -1,14 +1,33 @@
+ #!/bin/sh /etc/rc.common
+-# Copyright (C) 2008-2013 OpenWrt.org
++# Copyright (C) 2008-2015 OpenWrt.org
+
+ . $IPKG_INSTROOT/lib/functions/olsrd.sh
+
+ START=65
++USE_PROCD=1
++BIN=/usr/sbin/olsrd
++OLSRD=olsrd
++CONF=/var/etc/$OLSRD.conf
+
+-SERVICE_DAEMONIZE=1
+-SERVICE_WRITE_PID=1
++start_service() {
++	olsrd_generate_config $OLSRD
+
+-CONF=/var/etc/olsrd.conf
+-PID=/var/run/olsrd.pid
++	procd_open_instance
+
+-UCI_CONF_NAME=olsrd
++	procd_set_param command "$BIN"
++	procd_append_param command -f ${CONF}
++	procd_append_param command -nofork
++
++	procd_set_param respawn
++
++	procd_close_instance
++
++	olsrd_setup_smartgw_rules
++}
++
++service_triggers() {
++	procd_add_reload_trigger $OLSRD
```

Comment:
I think restarting olsrd if the uci config changes is not a good default behaviour. We may should restart if uci network changes but this should be covered with `procd_add_reload_interface_trigger`